Type size mismatch when reading compound datatype

Dear developers,

I have experienced an error "_LoadError: AssertionError: Type sizes mismatch!_" when trying to read a complicated compound datatype containing a integers, reals, and string arrays produced by our simulation code. From what I could find out, this issue seems to be caused by the size of the HDF5.FixedArray{String,(1,),1} datatype not being consistent with the size of the string array in the HDF5 file.
In the attachment, I have provided a small example (a Python script to create the HDF5 file and a Julia script to read it) where the compound datatype just contains a string array. With Julia 1.4.2 and a freshly installed HDF5.jl, it produces the following output:

`ERROR: LoadError: AssertionError: Type sizes mismatch!
Stacktrace:
 [1] read(::HDF5Dataset, ::Type{Array{NamedTuple{(:name,),Tuple{HDF5.FixedArray{String,(1,),1}}},N} where N}) at /home/<username>/.julia/packages/HDF5/cvKpY/src/HDF5.jl:1490
 [2] read(::HDF5Dataset) at /home/<username>/.julia/packages/HDF5/cvKpY/src/HDF5.jl:1306
 [3] read(::HDF5File, ::String, ::Bool) at /home/<username>/.julia/packages/HDF5/cvKpY/src/HDF5.jl:1294
 [4] read(::HDF5File, ::String) at /home/<username>/.julia/packages/HDF5/cvKpY/src/HDF5.jl:1293
 [5] top-level scope at /home/<username>/tests/test_hdf5_jl_bug/read_file.jl:5
 [6] include(::Module, ::String) at ./Base.jl:377
 [7] exec_options(::Base.JLOptions) at ./client.jl:288
 [8] _start() at ./client.jl:484
in expression starting at /home/<username>/tests/test_hdf5_jl_bug/read_file.jl:5
`

I hope that this example helps to figure out the issue. Another observation I made is that with Julia 1.2 and an older version of HDF5.jl, the String arrays are recognized as UInt8 and the compound datatype can be read successfully (however, one needs to decode the String manually).

Thank you for this very useful library.

[example.zip](https://github.com/JuliaIO/HDF5.jl/files/5083647/example.zip)

--- a/src/HDF5.jl
+++ b/src/HDF5.jl
@@ -2460,16 +2460,6 @@ function vlen_get_buf_size(dset::HDF5Dataset, dtype::HDF5Datatype, dspace::HDF5D
     sz[]
 end
 
-function hdf5array(objtype)
-    nd = h5t_get_array_ndims(objtype.id)
-    dims = Vector{Hsize}(undef,nd)
-    h5t_get_array_dims(objtype.id, dims)
-    eltyp = HDF5Datatype(h5t_get_super(objtype.id))
-    T = get_mem_compatible_jl_type(objtype)
-    dimsizes = ntuple(i -> Int(dims[nd-i+1]), nd)  # reverse order
-    FixedArray{T, dimsizes, prod(dimsizes)}
-end
-
 ### Property manipulation ###
 get_create_properties(d::HDF5Dataset)   = HDF5Properties(h5d_get_create_plist(d.id), H5P_DATASET_CREATE)
 get_create_properties(g::HDF5Group)     = HDF5Properties(h5g_get_create_plist(g.id), H5P_GROUP_CREATE)

--- a/src/HDF5.jl
+++ b/src/HDF5.jl
@@ -2376,6 +2376,18 @@ function h5l_get_info(link_loc_id::Hid, link_name::String, lapl_id::Hid)
     info[]
 end
 
+function h5lt_dtype_to_text(dtype_id; max_len=512)
+  buf = Vector{UInt8}(undef, max_len)
+  sz = Csize_t[length(buf)]
+  status = ccall((:H5LTdtype_to_text, libhdf5_hl), Int, (HDF5.Hid, Ptr{UInt8}, Int, Ptr{Csize_t}), dtype_id, buf, 0, sz)
+  if status < 0
+    error("Error getting dtype text representation")
+  end
+  buf[end] = 0 # nullterm
+  dtype_text = rstrip(unsafe_string(pointer(buf)))
+  return dtype_text
+end
+
 # Set MPIO properties in HDF5.
 # Note: HDF5 creates a COPY of the comm and info objects.
 function h5p_set_fapl_mpio(fapl_id, comm, info)

--- a/src/HDF5.jl
+++ b/src/HDF5.jl
@@ -1491,7 +1491,7 @@ function read(dset::HDF5Dataset, T::Union{Type{Array{U}}, Type{U}}) where U <: N
     h5type_str = h5lt_dtype_to_text(memtype.id)
     error("""
           Type size mismatch
-          sizeof($T) = $(sizeof(T))
+          sizeof($U) = $(sizeof(U))
           sizeof($h5type_str) = $(h5t_get_size(memtype.id))
           """)
   end

--- a/src/HDF5.jl
+++ b/src/HDF5.jl
@@ -479,7 +479,7 @@ eltype(::Type{FixedArray{T,D,L}}) where {T,D,L} = T
 eltype(x::T) where T <: FixedArray = eltype(T)
 
 struct FixedString{N, PAD}
-  data::NTuple{N, Cchar}
+  data::NTuple{N, UInt8}
 end
 length(::Type{FixedString{N, PAD}}) where {N, PAD} = N
 length(x::T) where T <: FixedString = length(T)
@@ -1470,7 +1470,7 @@ end
 normalize_types(x) = x
 normalize_types(x::NamedTuple{T}) where T = NamedTuple{T}(map(normalize_types, values(x)))
 normalize_types(x::Cstring) = unsafe_string(x)
-normalize_types(x::FixedString) = unpad(join(Char.(x.data)), pad(x))
+normalize_types(x::FixedString) = unpad(String(collect(x.data)), pad(x))
 normalize_types(x::FixedArray) = reshape(collect(x.data), size(x)...)
 normalize_types(x::VariableArray) = copy(unsafe_wrap(Array, convert(Ptr{eltype(x)}, x.p), x.len, own=false))
 

--- a/src/HDF5.jl
+++ b/src/HDF5.jl
@@ -2386,15 +2386,16 @@ function h5l_get_info(link_loc_id::Hid, link_name::String, lapl_id::Hid)
     info[]
 end
 
-function h5lt_dtype_to_text(dtype_id; max_len=512)
-  buf = Vector{UInt8}(undef, max_len)
-  sz = Csize_t[length(buf)]
-  status = ccall((:H5LTdtype_to_text, libhdf5_hl), Int, (HDF5.Hid, Ptr{UInt8}, Int, Ptr{Csize_t}), dtype_id, buf, 0, sz)
-  if status < 0
-    error("Error getting dtype text representation")
-  end
-  buf[end] = 0 # nullterm
-  dtype_text = rstrip(unsafe_string(pointer(buf)))
+function h5lt_dtype_to_text(dtype_id)
+  len = Csize_t[0]
+  status = ccall((:H5LTdtype_to_text, libhdf5_hl), Int, (HDF5.Hid, Ptr{UInt8}, Int, Ptr{Csize_t}), dtype_id, C_NULL, 0, len)
+  status < 0 && error("Error getting dtype text representation")
+
+  buf = Vector{UInt8}(undef, len[1])
+  status = ccall((:H5LTdtype_to_text, libhdf5_hl), Int, (HDF5.Hid, Ptr{UInt8}, Int, Ptr{Csize_t}), dtype_id, buf, 0, len)
+  status < 0 && error("Error getting dtype text representation")
+
+  dtype_text = unsafe_string(pointer(buf))
   return dtype_text
 end
 

--- a/src/HDF5.jl
+++ b/src/HDF5.jl
@@ -1484,21 +1484,31 @@ do_reclaim(::Type{T}) where T <: Union{Cstring, VariableArray} = true
 
 function read(dset::HDF5Dataset, T::Union{Type{Array{U}}, Type{U}}) where U <: NamedTuple
   filetype = datatype(dset)
-  memtype_id = h5t_get_native_type(filetype.id)  # padded layout in memory
-  @assert sizeof(U) == h5t_get_size(memtype_id) "Type sizes mismatch!"
+  memtype = HDF5Datatype(h5t_get_native_type(filetype.id))  # padded layout in memory
+  close(filetype)
+
+  if sizeof(U) != h5t_get_size(memtype.id)
+    h5type_str = h5lt_dtype_to_text(memtype.id)
+    error("""
+          Type size mismatch
+          sizeof($T) = $(sizeof(T))
+          sizeof($h5type_str) = $(h5t_get_size(memtype.id))
+          """)
+  end
 
   buf = Array{U}(undef, size(dset))
+  memspace = dataspace(buf)
 
-  h5d_read(dset.id, memtype_id, H5S_ALL, H5S_ALL, H5P_DEFAULT, buf)
+  h5d_read(dset.id, memtype.id, memspace.id, HDF5.H5S_ALL, dset.xfer.id, buf)
   out = do_normalize(U) ? normalize_types.(buf) : buf
 
   if do_reclaim(U)
-    dspace = dataspace(dset)
     # NOTE I have seen this call fail but I cannot reproduce
-    h5d_vlen_reclaim(memtype_id, dspace.id, H5P_DEFAULT, buf)
+    h5d_vlen_reclaim(memtype.id, memspace.id, dset.xfer, buf)
   end
 
-  h5t_close(memtype_id)
+  close(memtype)
+  close(memspace)
 
   if T <: NamedTuple
     return out[1]

--- a/src/HDF5.jl
+++ b/src/HDF5.jl
@@ -1471,8 +1471,8 @@ normalize_types(x) = x
 normalize_types(x::NamedTuple{T}) where T = NamedTuple{T}(map(normalize_types, values(x)))
 normalize_types(x::Cstring) = unsafe_string(x)
 normalize_types(x::FixedString) = unpad(String(collect(x.data)), pad(x))
-normalize_types(x::FixedArray) = reshape(collect(x.data), size(x)...)
-normalize_types(x::VariableArray) = copy(unsafe_wrap(Array, convert(Ptr{eltype(x)}, x.p), x.len, own=false))
+normalize_types(x::FixedArray) = normalize_types.(reshape(collect(x.data), size(x)...))
+normalize_types(x::VariableArray) = normalize_types.(copy(unsafe_wrap(Array, convert(Ptr{eltype(x)}, x.p), x.len, own=false)))
 
 do_normalize(::Type{T}) where T = false
 do_normalize(::Type{NamedTuple{T, U}}) where T where U = any(i -> do_normalize(fieldtype(U,i)), 1:fieldcount(U))

--- a/src/HDF5.jl
+++ b/src/HDF5.jl
@@ -1499,7 +1499,7 @@ function read(dset::HDF5Dataset, T::Union{Type{Array{U}}, Type{U}}) where U <: N
   buf = Array{U}(undef, size(dset))
   memspace = dataspace(buf)
 
-  h5d_read(dset.id, memtype.id, memspace.id, HDF5.H5S_ALL, dset.xfer.id, buf)
+  h5d_read(dset.id, memtype.id, memspace.id, H5S_ALL, dset.xfer.id, buf)
   out = do_normalize(U) ? normalize_types.(buf) : buf
 
   if do_reclaim(U)
@@ -1976,7 +1976,7 @@ function hdf5_to_julia_eltype(objtype)
     else
         error("Class id ", class_id, " is not yet supported")
     end
-    T
+    return T
 end
 
 function get_mem_compatible_jl_type(objtype)
@@ -2387,16 +2387,16 @@ function h5l_get_info(link_loc_id::Hid, link_name::String, lapl_id::Hid)
 end
 
 function h5lt_dtype_to_text(dtype_id)
-  len = Csize_t[0]
-  status = ccall((:H5LTdtype_to_text, libhdf5_hl), Int, (HDF5.Hid, Ptr{UInt8}, Int, Ptr{Csize_t}), dtype_id, C_NULL, 0, len)
-  status < 0 && error("Error getting dtype text representation")
+    len = Ref{Csize_t}()
+    status = ccall((:H5LTdtype_to_text, libhdf5_hl), Herr, (Hid, Ptr{UInt8}, Int, Ref{Csize_t}), dtype_id, C_NULL, 0, len)
+    status < 0 && error("Error getting dtype text representation")
 
-  buf = Vector{UInt8}(undef, len[1])
-  status = ccall((:H5LTdtype_to_text, libhdf5_hl), Int, (HDF5.Hid, Ptr{UInt8}, Int, Ptr{Csize_t}), dtype_id, buf, 0, len)
-  status < 0 && error("Error getting dtype text representation")
+    buf = Vector{UInt8}(undef, len[])
+    status = ccall((:H5LTdtype_to_text, libhdf5_hl), Herr, (Hid, Ptr{UInt8}, Int, Ref{Csize_t}), dtype_id, buf, 0, len)
+    status < 0 && error("Error getting dtype text representation")
 
-  dtype_text = unsafe_string(pointer(buf))
-  return dtype_text
+    dtype_text = unsafe_string(pointer(buf))
+    return dtype_text
 end
 
 # Set MPIO properties in HDF5.

--- a/src/HDF5.jl
+++ b/src/HDF5.jl
@@ -1946,36 +1946,11 @@ function hdf5_to_julia_eltype(objtype)
             error("character set ", cset, " not recognized")
         end
     elseif class_id == H5T_INTEGER || class_id == H5T_FLOAT
-        native_type = h5t_get_native_type(objtype.id)
-        try
-            native_size = h5t_get_size(native_type)
-            if class_id == H5T_INTEGER
-                is_signed = h5t_get_sign(native_type)
-            else
-                is_signed = nothing
-            end
-            T = hdf5_type_map[(class_id, is_signed, native_size)]
-        finally
-            h5t_close(native_type)
-        end
+        T = get_mem_compatible_jl_type(objtype)
     elseif class_id == H5T_BITFIELD
-        T = Bool
-        type_id = h5t_copy(hdf5_type_id(T))
-        h5t_set_precision(type_id, 1)
+        T = get_mem_compatible_jl_type(objtype)
     elseif class_id == H5T_ENUM
-        super_type = h5t_get_super(objtype.id)
-        try
-            native_type = h5t_get_native_type(super_type)
-            try
-                native_size = h5t_get_size(native_type)
-                is_signed = h5t_get_sign(native_type)
-                T = hdf5_type_map[(H5T_INTEGER, is_signed, native_size)]
-            finally
-                h5t_close(native_type)
-            end
-        finally
-            h5t_close(super_type)
-        end
+        T = get_mem_compatible_jl_type(objtype)
     elseif class_id == H5T_REFERENCE
         # How to test whether it's a region reference or an object reference??
         T = HDF5ReferenceObj
@@ -1985,6 +1960,58 @@ function hdf5_to_julia_eltype(objtype)
         super_id = h5t_get_super(objtype.id)
         T = HDF5Vlen{hdf5_to_julia_eltype(HDF5Datatype(super_id))}
     elseif class_id == H5T_COMPOUND
+        T = get_mem_compatible_jl_type(objtype)
+    elseif class_id == H5T_ARRAY
+        T = get_mem_compatible_jl_type(objtype)
+    else
+        error("Class id ", class_id, " is not yet supported")
+    end
+    T
+end
+
+function get_mem_compatible_jl_type(objtype)
+    class_id = h5t_get_class(objtype.id)
+    if class_id == H5T_STRING
+        if h5t_is_variable_str(objtype.id)
+          return Cstring
+        else
+          n = h5t_get_size(objtype.id)
+          pad = h5t_get_strpad(objtype.id)
+          return FixedString{Int(n), pad}
+        end
+    elseif class_id == H5T_INTEGER || class_id == H5T_FLOAT
+        native_type = h5t_get_native_type(objtype.id)
+        try
+            native_size = h5t_get_size(native_type)
+            if class_id == H5T_INTEGER
+                is_signed = h5t_get_sign(native_type)
+            else
+                is_signed = nothing
+            end
+            return hdf5_type_map[(class_id, is_signed, native_size)]
+        finally
+            h5t_close(native_type)
+        end
+    elseif class_id == H5T_BITFIELD
+        return Bool
+    elseif class_id == H5T_ENUM
+        super_type = h5t_get_super(objtype.id)
+        try
+            native_type = h5t_get_native_type(super_type)
+            try
+                native_size = h5t_get_size(native_type)
+                is_signed = h5t_get_sign(native_type)
+                return hdf5_type_map[(H5T_INTEGER, is_signed, native_size)]
+            finally
+                h5t_close(native_type)
+            end
+        finally
+            h5t_close(super_type)
+        end
+    elseif class_id == H5T_VLEN
+        superid = h5t_get_super(objtype.id)
+        return VariableArray{get_mem_compatible_jl_type(HDF5Datatype(superid))}
+    elseif class_id == H5T_COMPOUND
         N = h5t_get_nmembers(objtype.id)
 
         membernames = ntuple(N) do i
@@ -1992,23 +2019,8 @@ function hdf5_to_julia_eltype(objtype)
         end
 
         membertypes = ntuple(N) do i
-            dtype = HDF5Datatype(h5t_get_member_type(objtype.id, i-1))
-            ci = h5t_get_class(dtype.id)
-
-            if ci == H5T_STRING
-                if h5t_is_variable_str(dtype.id)
-                    return Cstring
-                else
-                    n = h5t_get_size(dtype.id)
-                    pad = h5t_get_strpad(dtype.id)
-                    return FixedString{Int(n), pad}
-                end
-            elseif ci == H5T_VLEN
-                superid = h5t_get_super(dtype.id)
-                T = VariableArray{hdf5_to_julia_eltype(HDF5Datatype(superid))}
-            else
-                return hdf5_to_julia_eltype(dtype)
-            end
+          dtype = HDF5Datatype(h5t_get_member_type(objtype.id, i-1))
+          return get_mem_compatible_jl_type(dtype)
         end
 
         # check if should be interpreted as complex
@@ -2019,19 +2031,21 @@ function hdf5_to_julia_eltype(objtype)
                     (membertypes[1] <: HDF5Scalar)
 
         if iscomplex
-            T = Complex{membertypes[1]}
+          return Complex{membertypes[1]}
         else
-            T = NamedTuple{Symbol.(membernames), Tuple{membertypes...}}
+          return NamedTuple{Symbol.(membernames), Tuple{membertypes...}}
         end
     elseif class_id == H5T_ARRAY
-        T = hdf5array(objtype)
-    else
-        error("Class id ", class_id, " is not yet supported")
+        nd = h5t_get_array_ndims(objtype.id)
+        dims = Vector{Hsize}(undef,nd)
+        h5t_get_array_dims(objtype.id, dims)
+        eltyp = HDF5Datatype(h5t_get_super(objtype.id))
+        elT = get_mem_compatible_jl_type(eltyp)
+        dimsizes = ntuple(i -> Int(dims[nd-i+1]), nd)  # reverse order
+        return FixedArray{elT, dimsizes, prod(dimsizes)}
     end
-
-    return T
+    error("Class id ", class_id, " is not yet supported")
 end
-
 
 ### Convenience wrappers ###
 # These supply default values where possible
@@ -2428,7 +2442,7 @@ function hdf5array(objtype)
     dims = Vector{Hsize}(undef,nd)
     h5t_get_array_dims(objtype.id, dims)
     eltyp = HDF5Datatype(h5t_get_super(objtype.id))
-    T = hdf5_to_julia_eltype(eltyp)
+    T = get_mem_compatible_jl_type(objtype)
     dimsizes = ntuple(i -> Int(dims[nd-i+1]), nd)  # reverse order
     FixedArray{T, dimsizes, prod(dimsizes)}
 end

--- a/test/compound.jl
+++ b/test/compound.jl
@@ -1,6 +1,7 @@
 using Random, Test, HDF5
 
 import HDF5.datatype
+import Base.convert
 import Base.unsafe_convert
 
 struct foo
@@ -14,7 +15,7 @@ end
 struct foo_hdf5
   a::Float64
   b::Cstring
-  c::NTuple{20, Cchar}
+  c::NTuple{20, UInt8}
   d::NTuple{9, ComplexF64}
   e::HDF5.Hvl_t
 end
@@ -22,7 +23,7 @@ end
 function unsafe_convert(::Type{foo_hdf5}, x::foo)
   foo_hdf5(x.a,
            Base.unsafe_convert(Cstring, x.b),
-           ntuple(i -> i <= length(x.c) ? x.c[i] : '\0', 20),
+           ntuple(i -> i <= ncodeunits(x.c) ? codeunit(x.c, i) : '\0', 20),
            ntuple(i -> x.d[i], length(x.d)),
            HDF5.Hvl_t(length(x.e), pointer(x.e))
           )
@@ -38,7 +39,7 @@ function datatype(::Type{foo_hdf5})
   HDF5.h5t_insert(dtype, "b", fieldoffset(foo_hdf5, 2), vlenstr_dtype)
 
   fixedstr_dtype = HDF5.h5t_copy(HDF5.H5T_C_S1)
-  HDF5.h5t_set_size(fixedstr_dtype, 20 * sizeof(Cchar))
+  HDF5.h5t_set_size(fixedstr_dtype, 20 * sizeof(UInt8))
   HDF5.h5t_set_cset(fixedstr_dtype, HDF5.H5T_CSET_UTF8)
   HDF5.h5t_set_strpad(fixedstr_dtype, HDF5.H5T_STR_NULLPAD)
   HDF5.h5t_insert(dtype, "c", fieldoffset(foo_hdf5, 3), fixedstr_dtype)
@@ -53,6 +54,34 @@ function datatype(::Type{foo_hdf5})
   HDF5Datatype(dtype)
 end
 
+struct bar
+  a::Vector{String}
+end
+
+struct bar_hdf5
+  a::NTuple{2, NTuple{20, UInt8}}
+end
+
+function datatype(::Type{bar_hdf5})
+  dtype = HDF5.h5t_create(HDF5.H5T_COMPOUND, sizeof(bar_hdf5))
+
+  fixedstr_dtype = HDF5.h5t_copy(HDF5.H5T_C_S1)
+  HDF5.h5t_set_size(fixedstr_dtype, 20 * sizeof(UInt8))
+  HDF5.h5t_set_cset(fixedstr_dtype, HDF5.H5T_CSET_UTF8)
+
+  hsz = HDF5.Hsize[2]
+  array_dtype = HDF5.h5t_array_create(fixedstr_dtype, 1, hsz)
+
+  HDF5.h5t_insert(dtype, "a", fieldoffset(bar_hdf5, 1), array_dtype)
+
+  HDF5Datatype(dtype)
+end
+
+function convert(::Type{bar_hdf5}, x::bar)
+  bar_hdf5(ntuple(i -> ntuple(j -> j <= ncodeunits(x.a[i]) ? codeunit(x.a[i], j) : '\0', 20), 2))
+end
+
+
 @testset "compound" begin
   N = 10
   v = [foo(rand(),
@@ -62,20 +91,41 @@ end
            rand(1:10, rand(10:100))
           )
        for _ in 1:N]
+
+  v[1] = foo(1.0,
+             "uniçº∂e",
+             "uniçº∂e",
+             rand(ComplexF64, 3,3),
+             rand(1:10, rand(10:100)))
+
   v_write = unsafe_convert.(foo_hdf5, v)
+
+  w = [bar(["uniçº∂e", "hello"])]
+  w_write = convert.(bar_hdf5, w)
 
   fn = tempname()
   h5open(fn, "w") do h5f
-    dtype = datatype(foo_hdf5)
-    space = dataspace(v_write)
-    dset = HDF5.h5d_create(h5f.id, "data", dtype.id, space.id)
-    HDF5.h5d_write(dset, dtype.id, v_write)
+    foo_dtype = datatype(foo_hdf5)
+    foo_space = dataspace(v_write)
+    foo_dset = HDF5.h5d_create(h5f.id, "foo", foo_dtype.id, foo_space.id)
+    HDF5.h5d_write(foo_dset, foo_dtype.id, v_write)
+
+    bar_dtype = datatype(bar_hdf5)
+    bar_space = dataspace(w_write)
+    bar_dset = HDF5.h5d_create(h5f.id, "bar", bar_dtype.id, bar_space.id)
+    HDF5.h5d_write(bar_dset, bar_dtype.id, w_write)
   end
 
-  v_read = h5read(fn, "data")
+  v_read = h5read(fn, "foo")
   for field in (:a, :b, :c, :d, :e)
     f = x -> getfield(x, field)
     @test f.(v) == f.(v_read)
+  end
+
+  w_read = h5read(fn, "bar")
+  for field in (:a,)
+    f = x -> getfield(x, field)
+    @test f.(w) == f.(w_read)
   end
 
   T = NamedTuple{(:a, :b, :c, :d, :e, :f), Tuple{Int, Int, Int, Int, Int, Cstring}}

